:book: Proposal for ROSA networking

mzazrivec opened the pull request:
<!-- Thanks for this PR! If this is your first PR please read the [contributing guide](../CONTRIBUTING.md) -->
<!-- If this PR is still work-in-progress and is being open for visibility please prefix the title with `WIP:` -->

**What type of PR is this?**

/kind documentation

<!--
Add one of the following kinds:
/kind feature
/kind bug
/kind api-change
/kind cleanup
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind flake
/kind regression
/kind support
-->

**What this PR does / why we need it**:

This pull request contains a proposal for implementing provisioning of networking infrastructure for ROSA HCP clusters.

<!-- Enter a description of the change and why this change is needed -->

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Checklist**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR in which case these can be deleted 

 Please add an icon to the title of this PR, the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other) 
-->

- [ ] squashed commits
- [x] includes documentation
- [ ] includes emoji in title 
- [ ] adds unit tests
- [ ] adds or updates e2e tests

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note

```

mzazrivec committed:
Proposal for ROSA network

serngawy reviewed (changes requested):
Please add more details for the proposed CRD RosaNetwork. The proposal mention with `four attributes: name, AZ count, region and CIDR block for VPC` , what is the spec and status fields ?
- Is it namespace scope ? cluster scope ? 
- What is the status conditions ? ex; validate/phases/applied ?
- As you are using cloud formation, how do you propagate the errors from cloud formation template to the CR status ?
- ROSAControlPlane will refer to the RosaNetwork, What fields the RosaControlplane will read to figure out the subnet-id and zones
- How do you will handle deleting the RosaNetwork while there is a RosaControlPlane refer to it ?

mzazrivec commented:
> Please add more details for the proposed CRD RosaNetwork. The proposal mention with `four attributes: name, AZ count, region and CIDR block for VPC` , what is the spec and status fields ?
>

Yes, I added more details regarding the spec / status part of the new custom resource.

> * Is it namespace scope ? cluster scope ?

Namespaced. Added to the doc.

> * What is the status conditions ? ex; validate/phases/applied ?

This is the part I'm not 100% sure about, but I'd probably prefer being consistent with what CAPA already uses, e.g. something like:
```
status:
  conditions:
    - lastTransitionTime: "2025-03-19T16:29:24Z"
      status: "True"
      type: RosaNetworkReady
      message:
      reason:
```

> * As you are using cloud formation, how do you propagate the errors from cloud formation template to the CR status ?

Through the `status.resources` array. Added to the doc.

> * ROSAControlPlane will refer to the RosaNetwork, What fields the RosaControlplane will read to figure out the subnet-id and zones

Added to the doc.

> * How do you will handle deleting the RosaNetwork while there is a RosaControlPlane refer to it ?

On one hand it would make sense to prevent RosaNetwork deletion should it be referenced from an existing control plane, but not quite sure how to implement that without using some type of labeling or something else that would allow CAPA to see whether or not is the particular RosaNetwork being referenced or not. In short, I don't want to be implementing any checks in the initial implementation.

serngawy commented on docs/proposal/2025-02-24-rosa-networking.md:
```diff
@@ -0,0 +1,128 @@
+1. A new reconciler for the new custom resource, implementing creation and deletion. The reconciler will be using an existing [CloudFormation template from ROSA CLI](https://github.com/openshift/rosa/blob/master/cmd/create/network/templates/rosa-quickstart-default-vpc/cloudformation.yaml) and will use AWS CloudFormation API to create and delete the AWS CloudFormation stack.
+
+   Outputs and resources created in the cloudformation stack will be tracked under `status` of the `RosaNetwork` type. In particular, the `status` will have the following fields:
+   * publicSubnets: array of public subnet ids
```
make it one list of ;
- availabilityZone
  privateSubnet
  publicSubnet

mzazrivec replied:
I don't understand the above. Do you want all the AZs, private and public subnets to be rolled into one array?

mzazrivec replied:
Changed the `status` to be an array of the subnets grouped by AZs.

serngawy commented on docs/proposal/2025-02-24-rosa-networking.md:
```diff
@@ -0,0 +1,128 @@
+Adopting the CloudFormation template used by rosa-cli would mean that CAPA and the `RosaNetwork` custom resource would be relying on a mechanism that is know to work well and any changes or fixes implemented in ROSA CLI would be picked up automatically in CAPA.
+
+In practical terms, implementation of the proposal would mean:
+1. A new namespaced custom resource definition `RosaNetwork` in CAPA with four attributes: name, AZ count, region and CIRD block for VPC. These four attributes will become the `spec` part of the new `RosaNetwork` type.
```
please add the go struct OR CR example for the rosaNetwork.spec

serngawy commented on docs/proposal/2025-02-24-rosa-networking.md:
````diff
@@ -0,0 +1,128 @@
+   ```
+   status:
+     resources:
+       - resource: NATGateway1
````
Is this will duplication for subnet-ids ? may we don't need the list above for (zone, pub-subnet, priv-subnet) ?

mzazrivec replied:
Yes, duplicated, as a convenience function.

serngawy commented on docs/proposal/2025-02-24-rosa-networking.md:
````diff
@@ -0,0 +1,128 @@
+         status: CREATE_COMPLETE
+         reason:
+   ```
+   and will be reflecting the resource definitions from the CloudFormation template (`resource`) and the values coming from AWS CloudFormation API (`id`, `status` and `reason).
````
please add the go struct OR CR example for the rosaNetwork.status and include the suggested conditions

mzazrivec replied:
Added

serngawy commented on docs/proposal/20250224-rosa-networking.md:
```diff
@@ -0,0 +1,212 @@
+      name: rosa-network-01
+      namespace: default
+    spec:
+      availability_zone_count: 3
```
Is there a reason for not specifying the regions? which regions will be used if the count is 2 ?

serngawy replied:
Checking the template [here](https://github.com/openshift/rosa/blob/master/cmd/create/network/templates/rosa-quickstart-default-vpc/cloudformation.yaml#L7) can you clarify why is the max count is 3 ? some region like us-west-2 has 4 AZ.  
Lets have the availabilityZone as list of strings similar to [rosaControlPlane](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/controlplane/rosa/api/v1beta2/rosacontrolplane_types.go#L100) 
It should be okay if we will need to change the template a bit to specify the zones

mzazrivec replied:
> Is there a reason for not specifying the regions? which regions will be used if the count is 2 ?

It is possible to specify the region. It is one of the input parameters of the CF template.

I'm guessing you meant AZs here, not the regions. It's not possible to specify the AZs, because that's how the CF template is written. You just specify the number of AZs you want to be used and they are being taken in the alphabetical order.

mzazrivec replied:
> Checking the template [here](https://github.com/openshift/rosa/blob/master/cmd/create/network/templates/rosa-quickstart-default-vpc/cloudformation.yaml#L7) can you clarify why is the max count is 3 ? some region like us-west-2 has 4 AZ.

Don't know why 3 is the limit. But that's how the current template is.

> Lets have the availabilityZone as list of strings similar to [rosaControlPlane](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/controlplane/rosa/api/v1beta2/rosacontrolplane_types.go#L100) It should be okay if we will need to change the template a bit to specify the zones

That would be beyond the scope of this proposal document and the initial implementation. It's also not just a matter of changing the CF template. This change would mean that the AZs are the `spec` of the CR, not the `status` of the CR. Or maybe even both the `spec` and `status` depending on whether you want to specify the AZ count or the AZ list.

It could potentially be a future enhancement, though I'm not entirely convinced that it would all that useful. In the multi-zone cluster installation you're probably more interested in the number of zones you're going to be using rather than their names, which is how the template is written now.

hunterkepley replied:
I'm not sure what the end goal for the network model here is, if it's to be used with clusters like I assume, then 3 is the correct number for multi-AZ. You can either have 1 AZ for single-AZ cluster, or 3 for multi-AZ. There is validation in the API which only allows one of 2 options (1 or 3) for every cluster created:

```
// ValidateAvailabilityZonesCount is responsible for verifying whether the number of availability zones adheres to specific rules:
//
// * The number of availability zones for a multi AZ cluster should be 3.
```

This validation is performed for every cluster in which the user has selected AZs: `if isAvailabilityZonesSet || selectAvailabilityZones {`

mzazrivec replied:
> To let end user has the ability to set the AZs its not beyond the proposal scope.
> 
> * As endUser using us-west-2 region it has 4 AZs, Why I cannot create subnets in all AZs (forced to 3 AZs) ? Is there a technical reason for that ?

That particular problem / question is not really related to this proposal nor can I really address it here.

> * As endUser us-east-2 region, I want to specify which AZ that I can create subnet to it ? Is there technical reason that I cannot specify AZ ? (The RosaControlPlane has the option to specify the AZs and subnets)

Well, technical reason .... There's a technical limitation right now: the CF template in rosa-cli does not allow it / support it. Things may change on that front for sure and it may be an enhancement (meaning both in rosa-cli & CAPA) in the future, but right now it's not possible.

serngawy commented on docs/proposal/20250224-rosa-networking.md:
```diff
@@ -0,0 +1,212 @@
+    spec:
+      availability_zone_count: 3
+      region: us-west-2
+      cidr_block: 10.0.0.0/16
```
cidrBlock

mzazrivec replied:
Fixed.

serngawy commented on docs/proposal/20250224-rosa-networking.md:
```diff
@@ -0,0 +1,212 @@
+     name: rosa-network-01
+     namespace: default
+   status:
+     publicSubnets:
```
These lists cannot tell which public, private, zone belong to each other, better to have it as below 
```suggestion
     subnets:
     - availabilityZone: us-west-2a
       privateSubnet: subnet-1d9f28ba992a83514
       publicSubnet: subnet-0d9f28ba991b93514
     - availabilityZone: us-west-2b
       privateSubnet: subnet-1d9f28ba992a83513
       publicSubnet: subnet-0d9f28ba991b93513    
```

serngawy commented on docs/proposal/20250224-rosa-networking.md:
```diff
@@ -0,0 +1,212 @@
+       name: hcp01-control-plane
+       namespace: default
+   spec:
+     RosaNetworkRef:
```
rosaNetworkRef
```suggestion
     rosaNetworkRef:
```

serngawy commented on docs/proposal/20250224-rosa-networking.md:
````diff
@@ -0,0 +1,212 @@
+     RosaNetworkRef:
+       name: hcp01-rosa-network
+   ```
+   Should the ROSA control plane CR contain reference to ROSA network, the reconciler will read the AWS region, AZ and subnet ids parameters from the ROSA network CR.
````
Lets also mention that end user cannot set both RosaControlPlane->spec->[subnets](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/controlplane/rosa/api/v1beta2/rosacontrolplane_types.go#L95) and rosaNetworkRef at same time.

mzazrivec replied:
Technically, the user can specify both, but CAPA would prefer one to the other, right?

serngawy replied:
yes, we will need to validate that in the webook raising error if user set both.

mzazrivec replied:
Added to the doc.

serngawy commented on docs/proposal/20250224-rosa-networking.md:
```diff
@@ -0,0 +1,221 @@
+      name: rosa-network-01
+      namespace: default
+    spec:
+      availabilityZoneCount: 3
```
since we have the availabilityZones, no need for availabilityZoneCount,  what do you think ?

mzazrivec replied:
I think that as long the CF template we're using supports both scenarios, we can support both as well.

nrb replied:
Are the `availabilityZones` and `availibilityZoneCount` fields intended to be mutually exclusive?

Based on the two examples provided, I'm assuming they are, however the proposal should state this explicitly.

mzazrivec replied:
@nrb Those 2 parameters are mutually exclusive in the sense that explicit AZs take priority over the AZ count. But the CF template we'll be using is written so that if you specify both, the stack provisioning succeeds (with explicit AZs taking priority). I can mention this in this in the proposal doc as well.

serngawy commented:
Thanks @mzazrivec,  looks good. You may mention as well during ROSANetwork deletion a validation must be applied to check if there is a RosaControlplane refer to it.

nrb commented on docs/proposal/20250224-rosa-networking.md:
```diff
@@ -0,0 +1,221 @@
+
+#### Functional Requirements
+
+1. Ability to create a new namespaced custom resource `RosaNetwork` with four attributes: name, AZ count, region and CIDR block for VPC.
```
This should probably note the AZ list, too.

mzazrivec replied:
@nrb yep, added to the proposal doc.

nrb commented on docs/proposal/20250224-rosa-networking.md:
````diff
@@ -0,0 +1,221 @@
+       privateSubnet: subnet-7d7e19c0af1f4d57f
+   ```
+
+   All resources created in the cloudformation stack will be tracked under `status.resources` array:
````
What's the reason for setting `subnets` as a top-level field instead of listing them under `status.resources.subnets`?

mzazrivec replied:
@nrb in the end the subnets would be listed under both `status.subnets` (grouped by AZ) and `status.resources`. The `status.resources` will be a list filled with the data propagated from AWS as the stack creation progresses. My thought here was to duplicate the results we're interested in (subnets & AZs) so that we don't have to parse it out of the `status.resources` (i.e. a convenience function).

nrb replied:
So if I could rephrase to confirm I understand - the subnets are promoted to a top level resource primarily for ease of readability?

mzazrivec replied:
@nrb Yes, I guess you could say that. Although I meant it mostly for the benefit of readibility by other controllers: AZs and subnet ids is what the ROSA controller needs to provision the actual cluster.

Do you think this is not a good idea?

nrb commented on docs/proposal/20250224-rosa-networking.md:
```diff
@@ -0,0 +1,221 @@
+     conditions:
+     - lastTransitionTime: "2025-03-20T14:45:26Z"
+       status: "True"
+       type: RosaNetworkReady
```
How will the `true` status be calculated for this? That all resources are in `CREATE_COMPLETE` from the AWS API, or something else?

nrb replied:
I'll also mention the changes that happened in upstream CAPI with regards to the definition of `Ready`[1]. In this context, I believe it should mean "The network resources can pass traffic for workloads".

I don't think anything in this proposal is counter to what I just said, but I wanted to point that out for any other reviewers.

We should also try to adhere to the recommendations made in that CAEP. Since CAPA's conditions have not yet been updated, conditions defined by this CRD may not match the past conventions, but IMO we should use the new conventions when adding new resources.

[1] https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20240916-improve-status-in-CAPI-resources.md#readiness-and-availability

mzazrivec replied:
> How will the `true` status be calculated for this? That all resources are in `CREATE_COMPLETE` from the AWS API, or something else?

AWS API informs the user / consumer when the whole stack is `CREATE_COMPLETE`.

nrb commented:
I agree with the general direction this proposal takes. I made some comments around clarity and explicitness, paying special attention to `status.conditions` since there has been a significant effort upstream to align CAPI's usage with that of the Kubernetes project itself.

The examples given in YAML are very helpful! It may also be useful to include Go definitions in order to better understand optional or mutually exclusive fields, since these can often be excluded from YAML output due to the `omitEmpty` tag.

nrb commented on docs/proposal/20250224-rosa-networking.md:
```diff
@@ -0,0 +1,221 @@
+     conditions:
+     - lastTransitionTime: "2025-03-20T14:45:26Z"
+       status: "True"
+       type: RosaNetworkReady
```
If you can add the reason in here, I think that's the last of my comments.

mzazrivec replied:
@nrb yep, ack :smiley:

mzazrivec replied:
Not 100% sure if I understand the linked document correctly. The document says: `Use of the Reason field is required (currently in Cluster API reasons is added only when condition are false)`

Does that mean we can have an empty string and still show the `Reason` field? Or do we have to have a non-empty reason string even for success?

mzazrivec commented:
I updated the proposal document slightly (mainly filed names and such) to reflect the implementation in https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/5464

damdo commented:
/release-note-none

damdo commented:
/remove-lifecycle stale

mzazrivec commented:
/remove-lifecycle stale

damdo commented:
@mzazrivec what's the status of this one?

mzazrivec commented:
@damdo The status of this PR is the same as it has been for the past year -- I addressed comments & suggestions and now I'm waiting for someone for either merge it or close it (ideally with an explanation why). Nonetheless, the actual implementation for what this PR proposes has been merged long time ago.

richardcase commented:
As this is purely for ROSA i am going to:

/approve

And then leave it to others to lgtm and unhold

/hold

damdo reviewed (commented):
/lgtm

the main implementation merged for this proposal, so we can considered it done and keep this as documentation.

/unhold
